jerry-core/parser/js/js-parser-expr.c

```diff
@@ -2020,6 +2053,399 @@ parser_process_binary_opcodes (parser_context_t *context_p, /**< context */
+
+  if (current_token_type == LEXER_ASSIGN)
+  {
+    parser_branch_t skip_init;
+    lexer_next_token (context_p);
+    parser_emit_cbc_ext (context_p, CBC_EXT_DEFAULT_INITIALIZER);
+    parser_emit_cbc_forward_branch (context_p, CBC_BRANCH_IF_FALSE_FORWARD, &skip_init);
+    parser_emit_cbc (context_p, CBC_POP);
+
+    parser_parse_expression (context_p, PARSE_EXPR_NO_COMMA | PARSE_EXPR_LEFT_HAND_SIDE);
+    parser_set_branch_to_current_position (context_p, &skip_init);
+  }
+  else
+  {
+    context_p->token.type = current_token_type;
+  }
+
+  parser_process_binary_opcodes (context_p, 0);
+
+#if ENABLED (JERRY_DEBUGGER)
+  if ((JERRY_CONTEXT (debugger_flags) & JERRY_DEBUGGER_CONNECTED)
+      && ident_line_counter != context_p->last_breakpoint_line)
+  {
+    parser_emit_cbc (context_p, CBC_BREAKPOINT_DISABLED);
+    parser_flush_cbc (context_p);
+
+    parser_append_breakpoint_info (context_p, JERRY_DEBUGGER_BREAKPOINT_LIST, ident_line_counter);
+
+    context_p->last_breakpoint_line = ident_line_counter;
+  }
+#endif /* ENABLED (JERRY_DEBUGGER) */
+
+#if ENABLED (JERRY_LINE_INFO)
+  if (ident_line_counter != context_p->last_line_info_line)
+  {
+    parser_emit_line_info (context_p, ident_line_counter, false);
+  }
+#endif /* ENABLED (JERRY_LINE_INFO) */
+} /* parser_pattern_form_assignment */
+
+/**
+ * Parse pattern inside a pattern.
+ */
+static void
+parser_pattern_process_inner_pattern (parser_context_t *context_p, /**< context */
+                                      parser_pattern_flags_t flags, /**< flags */
+                                      uint16_t rhs_opcode, /**< opcode to process the rhs value */
+                                      uint16_t literal_index) /**< literal index for object pattern */
+{
+  JERRY_ASSERT (context_p->token.type == LEXER_LEFT_BRACE || context_p->token.type == LEXER_LEFT_SQUARE);
+
+  parser_pattern_flags_t options = (PARSER_PATTERN_INNER_PATTERN | (flags & PARSER_PATTERN_BINDING));
+
+  if (context_p->next_scanner_info_p->source_p == context_p->source_p)
+  {
+    if (rhs_opcode == CBC_EXT_REST_INITIALIZER)
+    {
+      parser_raise_error (context_p, PARSER_ERR_INVALID_DESTRUCTURING_PATTERN);
+    }
+    options |= PARSER_PATTERN_PARSE_RHS | PARSER_PATTERN_DEFAULT;
+  }
+
+  parser_pattern_emit_rsh (context_p, rhs_opcode, literal_index);
+
+  if (context_p->token.type == LEXER_LEFT_BRACE)
+  {
+    parser_parse_object_initializer (context_p, options);
+  }
+  else
+  {
+    parser_parse_array_initializer (context_p, options);
+  }
+
+  if (!(options & PARSER_PATTERN_DEFAULT))
+  {
+    lexer_next_token (context_p);
+  }
+
+  parser_emit_cbc (context_p, CBC_POP);
+} /* parser_pattern_process_inner_pattern */
+
+/**
+ * Parse array initializer.
+ */
+void
+parser_parse_array_initializer (parser_context_t *context_p, /**< context */
+                                parser_pattern_flags_t flags) /**< flags */
+{
+  JERRY_ASSERT (context_p->token.type == LEXER_LEFT_SQUARE);
```

Comment:
Would it be a good idea to have  a single `parser_parse_initializer` which checks the current token?

Reply:
Good idea!